src/compiler/checker.ts

```diff
@@ -18800,8 +18800,12 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 // If none of the type arguments for the outer type parameters contain type variables, it follows
                 // that the instantiated type doesn't reference type variables.
                 if (result.flags & TypeFlags.ObjectFlagsType && !((result as ObjectFlagsType).objectFlags & ObjectFlags.CouldContainTypeVariablesComputed)) {
-                    (result as ObjectFlagsType).objectFlags |= ObjectFlags.CouldContainTypeVariablesComputed |
-                        (some(typeArguments, couldContainTypeVariables) ? ObjectFlags.CouldContainTypeVariables : 0);
+                    const resultCouldContainTypeVariables = some(typeArguments, couldContainTypeVariables);
+                    // The above check may have caused the result's objectFlags to update if the result is referenced via typeArguments.
+                    if (!((result as ObjectFlagsType).objectFlags & ObjectFlags.CouldContainTypeVariablesComputed)) {
+                        (result as ObjectFlagsType).objectFlags |= ObjectFlags.CouldContainTypeVariablesComputed |
+                            (resultCouldContainTypeVariables ? ObjectFlags.CouldContainTypeVariables : 0);
+                    }
                 }

```

Comment:
This could also have been:

```ts
if ((result as ObjectFlagsType).objectFlags & ObjectFlags.CouldContainTypeVariablesComputed) {
    resultCouldContainTypeVariables &&= !!((result as ObjectFlagsType).objectFlags & ObjectFlags.CouldContainTypeVariables);
}
(result as ObjectFlagsType).objectFlags |= ObjectFlags.CouldContainTypeVariablesComputed |
    (resultCouldContainTypeVariables ? ObjectFlags.CouldContainTypeVariables : 0);
```

e.g., merge the result. But that felt wrong. But, my current fix doesn't feel all correct either. I'm going to investigate and see if there's a specific class of types that we're hitting.

Reply:
I also have a version which does something... clever. Grab `typeCount` before obtaining the result, and if its `id` is greater than the saved `typeCount`, then you know it's a new type and we can do the calculation. This stems from my thought that this optimization can only possibly be valid on newly created types.

But, that feels really unusual.

Reply:
Honestly, with that latter solution (given guidance on how to determine what a "new type" is), we could simplify things a bit and just force calculate the value unconditionally, dropping the change in instantiateAnonymousType.